File: backend/app/schemas/dataset.py
from typing import Literal

from pydantic import BaseModel, Field, model_validator


Direction = Literal["UP", "DOWN"]


class Station(BaseModel):
    id: str
    code: str
    name: str
    sequence: int
    mileage_km: float


class Track(BaseModel):
    id: str
    station_id: str
    code: str
    type: str = "ARRIVAL_DEPARTURE"
    direction: Direction | Literal["BOTH"] = "BOTH"
    capacity: int = 1


class Section(BaseModel):
    id: str
    from_station_id: str
    to_station_id: str
    direction: Direction
    min_run_sec: int = Field(gt=0)
    min_headway_sec: int = Field(gt=0)


class StopPlan(BaseModel):
    station_id: str
    arr_sec: int | None = Field(default=None, ge=0)
    dep_sec: int | None = Field(default=None, ge=0)
    track_id: str | None = None
    stop_type: Literal["ORIGIN", "STOP", "PASS", "DESTINATION"] = "STOP"

    @model_validator(mode="after")
    def validate_times(self) -> "StopPlan":
        if self.arr_sec is None and self.dep_sec is None:
            raise ValueError("到达和出发时刻不能同时为空")
        if self.arr_sec is not None and self.dep_sec is not None and self.dep_sec < self.arr_sec:
            raise ValueError("出发时刻不能早于到达时刻；跨日时刻请转换为次日秒数")
        return self


class Train(BaseModel):
    id: str
    train_no: str
    category: str
    direction: Direction
    priority: int = Field(ge=0, le=100)
    stops: list[StopPlan]


class Dataset(BaseModel):
    id: str
    name: str
    service_date: str
    source_type: Literal["SAMPLE", "IMPORTED"] = "SAMPLE"
    status: Literal["READY", "DRAFT"] = "READY"
    created_at: str
    stations: list[Station]
    tracks: list[Track]
    sections: list[Section]
    trains: list[Train]
    disclaimer: str


class DatasetSummary(BaseModel):
    id: str
    name: str
    service_date: str
    status: str
    train_count: int
    station_count: int
    created_at: str


class SampleDatasetRequest(BaseModel):
    sample_key: str = "demo-central-01"

File: frontend/src/types/dataset.ts
export interface Station {
  id: string
  code: string
  name: string
  sequence: number
  mileage_km: number
}

export interface Track {
  id: string
  station_id: string
  code: string
  direction: 'UP' | 'DOWN' | 'BOTH'
}

export interface StopPlan {
  station_id: string
  arr_sec: number | null
  dep_sec: number | null
  track_id: string | null
  stop_type: string
}

export interface Train {
  id: string
  train_no: string
  category: string
  direction: 'UP' | 'DOWN'
  priority: number
  stops: StopPlan[]
}

export interface Dataset {
  id: string
  name: string
  service_date: string
  status: string
  created_at: string
  disclaimer: string
  stations: Station[]
  tracks: Track[]
  trains: Train[]
}

export interface DatasetSummary {
  id: string
  name: string
  service_date: string
  status: string
  train_count: number
  station_count: number
  created_at: string
}

export interface Occupancy {
  resource_id: string
  resource_code: string
  train_id: string
  train_no: string
  start_sec: number
  end_sec: number
  planned_arr_sec: number
  planned_dep_sec: number
}

export interface DiagramPoint {
  station_id: string
  station_name: string
  mileage_km: number
  arr_sec: number | null
  dep_sec: number | null
}

export interface DiagramTrain {
  train_id: string
  train_no: string
  direction: string
  category: string
  points: DiagramPoint[]
}

export interface DiagramData {
  dataset_id: string
  stations: Station[]
  trains: DiagramTrain[]
  occupancies: Occupancy[]
}

